### ISSUE TITLE
[Feature] Add Multi-Token Prediction module

### ISSUE THREAD

**Issue by lessw2020** [pytorch/torchtitan#933]
We need to create a multi-token prediction module (same as what DeepSeek v3 used) that can plug into architectures such as DeepSeek v3 etc. and allow for multi-token prediction training with an arbitrary level of blocks.  (default =4). 
This should be a modular architecture in that it clones the existing block and appends to the neck of the model and auto-integrates for training.

**Comment by janEbert** [comment]
Hey! I've worked on this in https://github.com/janEbert/torchtitan/commit/8571a7fba098e1d3a8652c8cc59e728b1a182fbe.
It does not yet correctly support pipelining: the shared embedding and unembedding layers' gradients need to be manually all-reduced because the PyTorch pipelining APIs don't have explicit support for shared parameters, to the best of my knowledge.
I'm also worried that FSDP2 may not handle the shared parameters correctly, this also requires testing.

I went with a limitation that only full MTP modules can be pipelined, i.e., always including an embedding, 2d×d linear projection, Transformer block, and unembedding.

If you are fine with an external contribution for this, I'd be happy for comments on the implementation and, of course, making it work with pipelining as well.

**Comment by lessw2020** [comment]
Hi @janEbert  - nice, thanks for the update and yes would definitely be happy to get an external PR from you on this. 
I have been busy with groupGEMM and haven't had time to start on this.
We can look at the pipelining aspect as would like MTP to be enabled with it. 
Look forward to working with you and your PR!

**Comment by janEbert** [comment]
Awesome, thanks! Will submit a PR in due time!

**Comment by JavaZeroo** [comment]
> Awesome, thanks! Will submit a PR in due time!

any update?

**Comment by janEbert** [comment]
Sorry, I never got around to it and am now not working with TorchTitan anymore. :(

**Comment by pretidav** [comment]
any update on this topic? is there any open roadmap about mtp? thx

**Comment by Matrix-Z97** [comment]
Hi, I've successfully integrated the MTP module for DeepSeek-V3 by reusing the original model modules.
It fully supports mainstream parallel strategies including FSDP, TP, EP, and CP.
Meanwhile, it also supports loading and saving HuggingFace weights with the MTP module included.
I'm wondering if the community is interested in this implementation. If there is interest, it's glad  for me to contribute our implementation code to TorchTitan.

**Comment by tianyu-l** [comment]
@Matrix-Z97 thanks, feel free to create a PR. I can help review.

**Comment by Matrix-Z97** [comment]
@tianyu-l thanks,  I've already submitted the PR , please help review it at your convenience.  #3392
